### ISSUE TITLE
[dart2js] possible bug: if condition gets ignored in for loop

### ISSUE THREAD

**Issue by georglu** [dart-lang/sdk#48465]
I think I found a bug in dart2js 2.16.1. It looks like the dart2js compiler ignores an if condition with a counter in a for loop when it should not. The dart code:
```
void main() {
  var counter = 30;
  for (var i = 0; i < 8; i++) {
    counter++;
    print('loop $i, counter = $counter');
    if (counter > 3) {
      print('counter > 3, reset counter to 0');
      counter = 0;
    }
  }
}
```
And here the generated JS code, the if condition is missing:
```
main() {
  var counter, i;
  for (counter = 30, i = 0; i < 8; ++i, counter = 0) {
    A.printString("loop " + i + ", counter = " + (counter + 1));
    A.printString("counter > 3, reset counter to 0");
  }
}
```
Output of dart function as expected:
```
dart test.dart
loop 0, counter = 31
counter > 3, reset counter to 0
loop 1, counter = 1
loop 2, counter = 2
loop 3, counter = 3
loop 4, counter = 4
counter > 3, reset counter to 0
loop 5, counter = 1
loop 6, counter = 2
loop 7, counter = 3
```
Output of generated JavaScript:
```
nodejs test.js
loop 0, counter = 31
counter > 3, reset counter to 0
loop 1, counter = 1
counter > 3, reset counter to 0
loop 2, counter = 1
counter > 3, reset counter to 0
loop 3, counter = 1
counter > 3, reset counter to 0
loop 4, counter = 1
counter > 3, reset counter to 0
loop 5, counter = 1
counter > 3, reset counter to 0
loop 6, counter = 1
counter > 3, reset counter to 0
loop 7, counter = 1
counter > 3, reset counter to 0
```
When the variable `counter` is initialized with a value smaller than 3, the bug doesn't occur.

**Comment by rakudrama** [comment]
Thanks for the concise report.
The bug appears to be in value range analysis.
After this phase, the condition `counter > 3` has become constant.
This appears to be a long-standing bug. I will investigate further.

**Comment by rakudrama** [comment]
Fixed by 404edd94351b46290f533a7f7ecdc744164ba2ac
